webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/configuration/ConfigurationRepository.scala

```diff
@@ -0,0 +1,132 @@
+import com.normation.rudder.domain.policies.DirectiveRId
+import com.normation.rudder.repository.FullActiveTechniqueCategory
+import com.normation.rudder.repository.RoDirectiveRepository
+import com.normation.rudder.repository.xml.GitParseActiveTechniqueLibrary
+import com.normation.rudder.repository.xml.GitParseTechniqueLibrary
+import zio._
+import zio.syntax._
+
+/*
+ * Easier te manage data
+ */
+final case class ActiveDirective(activeTechnique: ActiveTechnique, directive: Directive)
+
+/*
+ * This class is in charge of loading and updating configuration objects in rudder.
+ * All configuration objects are versionned and are associated with a pair of identifier:
+ * - their unique identifier (an uuid),
+ * - their version identifier (a git commit id, ie a sha1/sha256)
+ *
+ * The repository has a notion of "current version", which is commit id for head in default branch.
+ */
+trait ConfigurationRepository extends RoConfigurationRepository with WoConfigurationRepository
+
+trait RoConfigurationRepository {
+
+  /*
+   * Get a directive and its matching active technique for the given (id, version)
+   */
+  def getDirective(id: DirectiveRId): IOResult[Option[ActiveDirective]]
+  def getTechnique(id: TechniqueId): IOResult[Option[Technique]]
+
+  def getDirectiveLibrary(ids: Set[DirectiveRId]): IOResult[FullActiveTechniqueCategory]
+}
+
+trait WoConfigurationRepository {
+
+}
+
+/****************************************************************************************/
+
+class ConfigurationRepositoryImpl(
+    roDirectiveRepository      : RoDirectiveRepository
+  , techniqueRepository        : TechniqueRepository
+  , parseActiveTechniqueLibrary: GitParseActiveTechniqueLibrary
+  , parseTechniques            : GitParseTechniqueLibrary
+) extends ConfigurationRepository {
+
+  override def getDirective(id: DirectiveRId): IOResult[Option[ActiveDirective]] = {
+    (id.revId match {
+      case None | Some(GitVersion.defaultRev) =>
+        roDirectiveRepository.getActiveTechniqueAndDirective(id)
+      case Some(r)                            =>
+        parseActiveTechniqueLibrary.getDirective(id.id, r)
+    }).map( _.map{ case (at, d) => ActiveDirective(at, d)} )
+  }
+
+  override def getTechnique(id: TechniqueId): IOResult[Option[Technique]] = {
+    id.version.revId match {
+      case None | Some(GitVersion.defaultRev) =>
+        techniqueRepository.get(id).succeed
+      case Some(r)                            =>
+        parseTechniques.getTechnique(id.name, id.version.version, r)
+    }
+  }
+
+
+  def getDirectiveLibrary(ids: Set[DirectiveRId]): IOResult[FullActiveTechniqueCategory] = {
+    def nonDefaultRev(revId: Option[RevId]): Boolean = revId.isDefined && revId != Some(GitVersion.defaultRev)
+    val versionedDirectives = ids.filter(x => nonDefaultRev(x.revId))
+    for {
+      optDirs   <- ZIO.foreach(versionedDirectives.toList)(getDirective) // TODO: find a way to do that without N git treewalks
```

Comment:
I think that would need a new function in GitParseActiveTechniqueLibrary that group directive id by revision and do a git command for each revision to get corresponding  directives

ids.groupMap(_.revision)(_.id).flatMap( (rev, ids) => parseActiveTechniqueLibrary.getDirective(ids,rev) )

Reply:
Before that, I beliieve we will need to be able to "merge" revision of directives. For now, we have nothing to tell that a directive at two point in the git tree is the same (and so, that its revision are the same) if there is no modification on that directive between the two commits.